### ISSUE TITLE
jooq generating invalid merge upsert on postgres

### ISSUE THREAD

**Issue by toesterdahl** [jOOQ/jOOQ#15923]
### Expected behavior

That I can use aliases. 
That the id value passed is interpreted as a uuid. Currently the db read it a as a string. 

### Actual behavior

I am migrating an application from postgres 12 and jooq 3.15.4 (professional) to postgres 16 and jooq 3.18.7. I have some queries using 'onDuplicateKeyUpdate()' causing problems on migration, since jooq now generate a 'merge into' query that does not execute on postgres. Before the migration it generated a 'insert into .. on conflict .. do update ..' that worked. I wonder what is going on here. Is the 'merge into' syntax supported at all on postgres? 

### Steps to reproduce the problem

Since I needed to ask this question I created a mock example of my problem. Ironically this mock example generate the 'insert into .. on conflict' query, that works. 

```
        return DSL
            .insertInto(
                fruitTable,
                fruitTable.FRUIT_ID,
                fruitTable.NAME,
            )
            .values(
                mandate.mandateId,
                "Banana",
            )
            .onDuplicateKeyUpdate()
            .set(fruitTable.NAME, "Apple")
```

Result:
```
insert into "public"."fruit" as "fr" ("fruit_id", "name") values ('2f8a4b65-7e1e-49d4-8530-ae2b6bb08857', 'Banana') on conflict ("fruit_id") do update set "name" = 'Apple';

-- is OK!
```

However my original example result in a 'merge into', that have a problem with the alias, 

```
    return DSL
            .insertInto(
                mandateTable,
                mandateTable.MANDATE_ID,
                mandateTable.LAST_UPDATE,
                )
            .values(
                mandate.mandateId,
                LocalDateTime.now(zid),
            )
        .onDuplicateKeyUpdate()
        .set(mandateTable.LAST_UPDATE, LocalDateTime.now(zid))
```

```
-- Executing SQL:

merge into "public"."mandate" as "m" using (select '2f8a4b65-7e1e-49d4-8530-ae2b6bb08857', timestamp '2023-12-07 18:05:33.033756553') as "t" ("mandate_id", "last_update") on ("public"."mandate"."mandate_id" = "t"."mandate_id" or "public"."mandate"."portfolio_id" = null) when matched then update set "last_update" = timestamp '2023-12-07 18:05:33.033877463' when not matched then insert ("mandate_id", "last_update") values ("t"."mandate_id", "t"."last_update");

-- org.jooq.exception.DataAccessException:
-- SQL [merge into "public"."mandate" as "m" using (select '2f8a4b65-7e1e-49d4-8530-ae2b6bb08857', timestamp '2023-12-07 18:05:33.033756553') as "t" ("mandate_id", "last_update") on ("public"."mandate"."mandate_id" = "t"."mandate_id" or "public"."mandate"."portfolio_id" = null) when matched then update set "last_update" = timestamp '2023-12-07 18:05:33.033877463' when not matched then insert ("mandate_id", "last_update") values ("t"."mandate_id", "t"."last_update")]; Batch entry 1 merge into "public"."mandate" as "m" using (select '2f8a4b65-7e1e-49d4-8530-ae2b6bb08857', timestamp '2023-12-07 18:05:33.033756553') as "t" ("mandate_id", "last_update") on ("public"."mandate"."mandate_id" = "t"."mandate_id" or "public"."mandate"."portfolio_id" = null) when matched then update set "last_update" = timestamp '2023-12-07 18:05:33.033877463' when not matched then insert ("mandate_id", "last_update") values ("t"."mandate_id", "t"."last_update")
-- was aborted: ERROR: invalid reference to FROM-clause entry for table "mandate"
--  Hint: Perhaps you meant to reference the table alias "m".
```

If I remove the alias I get into the next problem, 

```
    return DSL
            .insertInto(
                Tables.MANDATE,
                Tables.MANDATE.MANDATE_ID,
                Tables.MANDATE.LAST_UPDATE,
                )
            .values(
                Tables.MANDATE.mandateId,
                LocalDateTime.now(zid),
            )
        .onDuplicateKeyUpdate()
        .set(Tables.MANDATE.LAST_UPDATE, LocalDateTime.now(zid))
```

```
-- Executing SQL:

merge into "public"."mandate" using (select '39d98a7c-39f2-4898-840d-90df7b22884b', timestamp '2023-12-07 17:57:27.106425382') as "t" ("mandate_id", "last_update") on ("public"."mandate"."mandate_id" = "t"."mandate_id" or "public"."mandate"."portfolio_id" = null) when matched then update set "last_update" = timestamp '2023-12-07 17:57:27.106553633' when not matched then insert ("mandate_id", "last_update") values ("t"."mandate_id", "t"."last_update");

-- Result: org.jooq.exception.DataAccessException:
-- SQL [merge into "public"."mandate" using (select '39d98a7c-39f2-4898-840d-90df7b22884b', timestamp '2023-12-07 17:57:27.106425382') as "t" ("mandate_id", "last_update") on ("public"."mandate"."mandate_id" = "t"."mandate_id" or "public"."mandate"."portfolio_id" = null) when matched then update set "last_update" = timestamp '2023-12-07 17:57:27.106553633' when not matched then insert ("mandate_id", "last_update") values ("t"."mandate_id", "t"."last_update")]; Batch entry 1 merge into "public"."mandate" using (select '39d98a7c-39f2-4898-840d-90df7b22884b', timestamp '2023-12-07 17:57:27.106425382') as "t" ("mandate_id", "last_update") on ("public"."mandate"."mandate_id" = "t"."mandate_id" or "public"."mandate"."portfolio_id" = null) when matched then update set "last_update" = timestamp '2023-12-07 17:57:27.106553633' when not matched then insert ("mandate_id", "last_update") values ("t"."mandate_id", "t"."last_update")
-- was aborted: ERROR: operator does not exist: uuid = text
--   Hint: No operator matches the given name and argument types. You might need to add explicit type casts.

```

### jOOQ Version

3.18.7

### Database product and version

postgres 15, postgres 16

### Java Version

Open JDK 17

### OS Version

Linux Ubuntu 22.04

### JDBC driver name and version (include name if unofficial driver)

postgresql 42.2.13

**Comment by toesterdahl** [comment]
I upgraded the jdbc driver to postgresql 42.7.0, but with same outcome.

**Comment by lukaseder** [comment]
Thanks a lot for your report. I'll have a look at this soon.

**Comment by samunstein** [comment]
I think I ran into the second problem too. This should hopefully be sufficient information to reproduce that one.

```sql
create table foo (
  id uuid primary key,
  group_id uuid not null,
  bar int,
  baz int,

  unique (id, group_id)
);
```

A snippet of scala code to illustrate the unexpected behaviour:
```scala
val record = new FooRecord().setId(UUID.randomUUID()).setGroupId(UUID.randomUUID())
val query  = DSL.insertInto(Tables.FOO).set(record).onDuplicateKeyUpdate().set(record)
// This query will throw the error if executed by hand for example by copy pasting into psql console
println(query.getSQL(ParamType.INLINED))
// This succeeds for some reason
query.execute()
// This throws the same error as the printed query
DSL.batch(query).execute() 
```

The printed, erroring query looks like this (line breaks added to make it easier to read): 
```sql 
merge into "test"."foo" 
using (select '4db50be1-a57d-4059-a459-78eac7297372', '7c3c51e7-277c-40fd-9e4e-b78044785e21') 
as "t" ("id", "group_id") 
on ("test"."foo"."id" = "t"."id" or ("test"."foo"."id" = "t"."id" and "test"."foo"."group_id" = "t"."group_id")) 
when matched then update set "id" = '4db50be1-a57d-4059-a459-78eac7297372', "group_id" = '7c3c51e7-277c-40fd-9e4e-b78044785e21' 
when not matched then insert ("id", "group_id") values ("t"."id", "t"."group_id");
```

and the error, when hand-executed in psql, is positioned at
```
ERROR:  operator does not exist: uuid = text
LINE 4: on ("test"."foo"."id" = "t"."id" or ("test"."foo"."id" = "t"...
                              ^
HINT:  No operator matches the given name and argument types. You might need to add explicit type casts.
```

If you, however getSQL with ParamType.INDEXED, you get a slightly different query, with some additional casts to uuid:

```sql
merge into "test"."foo" 
using (select cast(? as uuid), cast(? as uuid)) 
as "t" ("id", "group_id") 
on ("test"."foo"."id" = "t"."id" or ("test"."foo"."id" = "t"."id" and "test"."foo"."group_id" = "t"."group_id")) 
when matched then update set "id" = cast(? as uuid), "group_id" = cast(? as uuid) 
when not matched then insert ("id", "group_id") values ("t"."id", "t"."group_id")
```

* jOOQ version: 3.19.3
* postgresql driver version: 42.5.4

**Comment by lukaseder** [comment]
> EDIT: Tried again on jOOQ 3.19.9, and could not reproduce anymore. My problem seems to be fixed in 3.19.6

Thanks for the updated feedback, @samunstein. Yes, this issue got backported to 3.19.6:

- https://github.com/jOOQ/jOOQ/issues/16334#issuecomment-1968705362
